src/System/IO/BlockIO.hs

```diff
@@ -95,6 +99,22 @@ defaultIOCtxParams =
     ioctxConcurrencyLimit = 64 * 3
   }
 
+validateIOCtxParams :: IOCtxParams -> Maybe String
+validateIOCtxParams IOCtxParams{..} = ("IOCtxParams are invalid because " ++) <$>
+    if
+      | ioctxBatchSizeLimit <= 0
+      -> Just "the batch size limit is non-positive"
+      | ioctxBatchSizeLimit >= 2^(15 :: Int)
+      -> Just "the batch size limit is greater than or equal to 2^15"
+      | ioctxConcurrencyLimit <= 0
+      -> Just "the concurrency limit is non-positive"
+      | ioctxConcurrencyLimit >= 2^(16 :: Int)
+      -> Just "the concurrency limit is greater than or equal to 2^16"
+      | ioctxBatchSizeLimit > ioctxConcurrencyLimit
```

Comment:
I'm not sure where these numbers come from.

I *think* the io-uring maximum ring size is 4096, which is 2^12. So that'd be the maximum for both batch size (SQ) and concurrency limit (CQ).

Reply:
Yeah, I should have specified. I got to these numbers by testing when `liburing` would complain. And in more recent versions, there are also macros that define the upper bounds

Reply:
For example, in `setup.c` I could find 

```c
#define KERN_MAX_ENTRIES	32768
#define KERN_MAX_CQ_ENTRIES	(2 * KERN_MAX_ENTRIES)
```

The former is the bound for the SQ ring size, and the latter is the bound for the CQ ring size